x/auth/changepubkey/keeper/keeper.go

```diff
@@ -0,0 +1,150 @@
+// thus creating a stack of all previous keys for an address and the time intervals for which they were active.
+// This allows dapps and clients to easily query past keys for an account which may be useful for features
+// such as verifying timestamped off-chain signed messages."
+
+// Logger returns a module-specific logger.
+func (pk Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
+}
+
+// GetPubKeyHistory Returns the PubKey history of the account at address by time: involves current pubkey
+func (pk Keeper) GetPubKeyHistory(ctx sdk.Context, addr sdk.AccAddress) ([]types.PubKeyHistory, error) {
+	entries := []types.PubKeyHistory{}
+	if pk.ak.GetAccount(ctx, addr) == nil {
+		return entries, fmt.Errorf("account %s not found", addr.String())
+	}
+	iterator := pk.PubKeyHistoryIterator(ctx, addr)
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		entry := types.DecodeHistoricalEntry(iterator.Value())
+		entries = append(entries, entry)
+	}
+	currentEntry := pk.GetCurrentPubKeyEntry(ctx, addr)
+	entries = append(entries, currentEntry)
+	return entries, nil
+}
+
+// GetPubKeyHistoricalEntry Returns the PubKey historical entry at a specific time: involves current pubkey
+func (pk Keeper) GetPubKeyHistoricalEntry(ctx sdk.Context, addr sdk.AccAddress, time time.Time) types.PubKeyHistory {
+	iterator := pk.PubKeyHistoryIteratorAfter(ctx, addr, time)
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		entry := types.DecodeHistoricalEntry(iterator.Value())
+		if entry.EndTime.After(time) || entry.EndTime.Equal(time) { // TODO: is this inclusive?
+			return entry
+		}
+	}
+
+	return pk.GetCurrentPubKeyEntry(ctx, addr)
+}
+
+// GetLastPubKeyHistoricalEntry Returns the PubKey historical entry of last pubkey: does not involve current pubkey
+func (pk Keeper) GetLastPubKeyHistoricalEntry(ctx sdk.Context, addr sdk.AccAddress) types.PubKeyHistory {
+	iterator := pk.PubKeyHistoryReverseIterator(ctx, addr)
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		entry := types.DecodeHistoricalEntry(iterator.Value())
+		return entry
+	}
+	return types.PubKeyHistory{
+		StartTime: time.Time{},
+		EndTime:   time.Time{},
+	}
+}
+
+// GetCurrentPubKeyEntry Returns the PubKey entry of current time
+func (pk Keeper) GetCurrentPubKeyEntry(ctx sdk.Context, addr sdk.AccAddress) types.PubKeyHistory {
+	time := ctx.BlockTime() // TODO: ctx.BlockTime() is correct for endTime?
+	acc := pk.ak.GetAccount(ctx, addr)
+	lastEntry := pk.GetLastPubKeyHistoricalEntry(ctx, addr)
+	st := lastEntry.EndTime
+	if st.Equal(time) {
+		st = lastEntry.GetStartTime()
+	}
+	var pubkey crypto.PubKey
+	if acc != nil {
+		pubkey = acc.GetPubKey()
+	}
+
+	return types.PubKeyHistory{
+		PubKey:    types.EncodePubKey(pk.cdc, pubkey),
+		StartTime: st,
+		EndTime:   time,
```

Comment:
Could we have something else to show that it's "still active" instead of just the current time?  Maybe just some sort of null time?